Kernel/SanCov.cpp

```diff
@@ -4,34 +4,78 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include <Kernel/Arch/Processor.h>
 #include <Kernel/Devices/KCOVDevice.h>
-#include <Kernel/Tasks/Thread.h>
+#include <Kernel/Library/Panic.h>
 
 extern bool g_in_early_boot;
 
-extern "C" {
+#ifdef ENABLE_KERNEL_COVERAGE_COLLECTION_DEBUG
+// Set kcov_emergency_off=true before making calls from __sanitizer_cov_trace_pc
+// to kcov instrumented code to prevent an infinite recursion. We don't want to
+// stop instrumenting all the code transitively called code used by
+// __sanitizer_cov_trace_pc during error handling. Not instrumented means we can't
+// fuzz it meaningfully.
+bool kcov_emergency_off { false };
 
-void __sanitizer_cov_trace_pc(void);
-void __sanitizer_cov_trace_pc(void)
+[[noreturn]] void crash_and_burn(Thread* thread);
+void crash_and_burn(Thread* thread)
+{
+    kcov_emergency_off = true;
+    thread->print_backtrace();
+    PANIC("KCOV is b0rked.");
+    VERIFY_NOT_REACHED();
+}
+
+#    define RETURN                                 \
```

Comment:
I'm really not a big fan of this. We generally don't use macro for this type of stuff, and if maintainers are ok to go with macro, we should name it with an explicit name, not just `RETURN`.
I'm not at the computer to test it and propose some code but you should be able to get away with a `ScopeGuard` and some templates.

Reply:
Using TemporaryChange instead, see ADKasters comment on this.